Caching mechanism to set and get latest records

eddowh opened the pull request:
**CHANGES**:
- Renamed `openkongqi/cache.py` to `openkongi/filecache.py`. This is to avoid confusion from the cache database that will store the latest records.
- Renamed configuration variable `CACHE` to `RESOURCE_CACHE` to explicitly indicate this is NOT the cache database.
- Add `cache` key in dictionary of config variable `DATABASES` to set up cache database for latest records.
- Created `openkongqi.cache` module for cache database wrappers.
- Initializer methods for `openkongqi.records` now requires a `openkongqi.cache.CacheWrapper` object as an argument.
- Changed terminology in documentation of `openkongqi.records` wrapper classes as to NOT assume the records db is a traditional relational database (i.e. InfluxDB)
- Minor fixes in `openkongqi.status` module.

**NOTE**: the actual logic behind `set_latest` and `get_latest` has not yet been implemented in the submodules of `openkongqi.records`

On top of #20 

Covers #23

Commit:
Stubs for caching db mechanism for latest records

NOTE: Renamed `openkongqi.cache` to `openkongqi.filecache` to avoid
confusion between caching resources and cache database for latest
records.

Minor changes
=============
- Accordingly changed configuration variable names.
- Fixed documentation here and there.
- Rename modules for consistency.

eddowh commented:
Includes tests for `openkongqi.records.sqlalchemy.get_records`, as requested in #20

hrbonz commented on tests/test_records_sqlalch.py:
```diff
@@ -0,0 +1,50 @@
+
+    def setUp(self):
+        db_settings = {
+            "NAME": "",  # in memory
```
make an explicit call to `:memory:`

hrbonz commented on tests/test_records_sqlalch.py:
```diff
@@ -0,0 +1,50 @@
+        cache = CacheWrapper({})
+        self.wrapper = RecordsWrapper(db_settings, cache)
+        self.wrapper.db_init()
+
```
Add a `tearDown` method to clean db when finished.

hrbonz commented on openkongqi/records/sqlalch.py:
```diff
@@ -64,14 +64,31 @@ def write_records(self, records):
+        query = self._cnx.query(Record).filter(Record.uuid == uuid)
+
+        # make input time parameters UTC or force it
+        def to_utc(dt):
```
Move this outside of the function

hrbonz commented on openkongqi/records/sqlalch.py:
```diff
@@ -64,14 +64,31 @@ def write_records(self, records):
+        for rec in query.all():
+            yield Record.to_tuple(rec)
+
+    def set_latest(self, uuid, record):
```
Set the redis caching here.

Commit:
Fix false negatives of empty content when fetching

Use HTTPretty for mock testing http requests

Commit:
Change data extraction format

{
  "cn:shanghai:jinganjiancezhan": [
    {
      "ts": datetime.datetime(2016, 7, 11, 16, 31, 10, tzinfo=<...>),
      "fields": {
        "pm25": 0,
        "pm10": 0,
        ...
      },
    },
    {...}
  ],
  "cn:shanghai:hongkou": [...],
  ...,
}

eddowh commented:
Replaced stubs with actual implementation.

Covers #23, #30

Tests as requested by #20

eddowh commented on openkongqi/records/sqlalch.py:
```diff
@@ -38,40 +38,86 @@ def create_cnx(self, settings):
+        ).count()
+        return dup_count != 0
+
+    def write_records(self, data):
```
Judging by the length of this function I can expect you to tell me to atomize it.

Here's what I think I can do:

Have a `write_record(uuid, record)` method where
- `uuid` is a uuid
- record is a dictionary with `ts` and `fields`

Have the complementary `write_records(uuid, records)` where:
- `uuid` is a uuid
- `records` is a list of dict

Now, due to the nature of the data scraped, we can get multiple UUIDs in one round, so I propose we actually have the main method to named `dump_data(data)` where:
- `data` will be a dictionary where key:pair == uuid:records

It won't take in UUIDs and it will just loop through each UUID and call `write_records`.

hrbonz commented on .gitignore:
```diff
@@ -12,3 +12,6 @@ docs/_build
 *.egg-info
+
+# dev
+celeryconfig_dev.py
```
This is your own celery config, add it to your `.git/info/exclude`

hrbonz commented on openkongqi/records/base.py:
```diff
@@ -74,7 +72,79 @@ def get_records(self, start, end, filters=None):
         """
         raise NotImplementedError
 
+    def _apply_key_ctx(self, key):
```
This is not necessary

hrbonz commented on openkongqi/records/base.py:
```diff
@@ -74,7 +72,79 @@ def get_records(self, start, end, filters=None):
+        else:
+            return key
+
+    def get_cache_key(self, uuid, contextualized=True):
```
This should look like:

``` py
def get_cache_key(self, uuid):
    context = {"uuid": uuid}
    if self.key_ctx is not None:
        context.update(self.key_ctx)
    return self._cache_key.format(**context)
```

Then my key should look like: `okq:{modname}:{uuid}:latest`

hrbonz commented:
After our discussion, points to address:
- add a static key context at the source level

``` py
class Source(HTTPSource):
    key_context = {...}
[...]
```
- pass this key context from `save_data()` down to `recsdb.get_cache_key()` this is ugly but it works
- create a `get_latest()` method on the source and pass the context too

This is in addition to the other comments I made on the code.

Commit:
Implement cache latest value when writing records

hrbonz commented:
A few more modifications:
- rename `modname` to `moduuid` to avoid confusion in naming things unrelated
- move `_get_rec_uuid` to the sqlalch module
- make `key_context` a public attribute and set it in the class as mentioned in the previous comment

Commit:
Change data format and test method for get_records

hrbonz commented:
Merged in f3d2790
